### ISSUE TITLE
VoiceOver & accessibility support

### ISSUE THREAD

**Issue by liorbentov** [JedWatson/react-select#1084]
Hey, 
I have to say that I love this component and use it often.
However, the accessibility support isn't completed - It doesn't support Screen reader (for my knowledge)
I'm using Mac's VoiceOver, and when i'm navigating between the options with the keyboard, I expect to hear the option text (or whatever value I can pass), but nothing happens.

Would love if the component will support this feature.

**Comment by jesperzach** [comment]
+1

**Comment by LuckXP** [comment]
+1

**Comment by LarsFinley** [comment]
It does need this to be fully accessible.

**Comment by zfalen** [comment]
+1

**Comment by sattaman** [comment]
+1

**Comment by rnagavel** [comment]
+1

**Comment by rnagavel** [comment]
With Voice over and Talkback screen readers on,  user is unable to select the option from the list by double tap.

**Comment by paperview** [comment]
+1

**Comment by roborourke** [comment]
@liorbentov I just went over this and it seems to work with Voiceover now for the most part. May be because an update to OSX.

It is a bit confusing, when you tab in it announces it as a group and you can use the arrow keys, each option is announced as the label and 'clickable'. When you tab out it just says 'collapsed, button'. Not great but better than no voiceover at all. If you have some specific bugs / improvements based on how it behaves now I can dig into it as I need it for work and must meet WCAG AA standards.

**Comment by paperview** [comment]
@roborourke Have you tried it with the Voiceover on iOS?  It hasn't been working for me at all in that use case.  Not sure if that is a part of the standards or not.

**Comment by reneedudley** [comment]
@roborourke voiceover on a webview in iOS is not working for me either. I am unable to select an option from the group. I have tried removing the fast click as well as preventing default behavior on touchStart. Attached is a video. Any help or insight would be great! 
https://imgur.com/a/VU2FQ

**Comment by roborourke** [comment]
Ah sorry, this was chrome on macOS. I’ll do some actually thorough testing

**Comment by reneedudley** [comment]
@roborourke - I've done some aria research which might be of help. Several people have documented that they are unable to select when role combobox is being used.  So a potential fix could be allowing the user to override the `role: 'combobox'` in the renderInput function and the `role: option` in the option render function. 
https://www.levelaccess.com/mobile-browser-support-for-aria-roles-states-and-properties/

**Comment by anthonyk1225** [comment]
Can confirm that NVDA reads out each option when navigating with arrow keys on IE & Firefox. macOs voiceover is silent on arrow key navigation using any browser.

**Comment by anthonyk1225** [comment]
I had to come up with a solution for VoiceOver and so I've made a PR - https://github.com/JedWatson/react-select/pull/2422. Let me know if you guys have any questions or insight!

**Comment by stonebat** [comment]
I'm having the same issue.  This content was generated by the react-select 1.2.1.  Somehow VO can't read the aria-label on the div with role="option".  Then I copied the <div class="Select-menu-outer"> tag to jsfiddle, and it's working.  https://jsfiddle.net/stonebat/w4xxas99/2/

Any one...any idea?

```
<div class="Select has-value is-clearable is-searchable Select--single">
  <input type="hidden" name="selected-state" value="south-australia">
  <div class="Select-control">
    <span class="Select-multi-value-wrapper" id="react-select-2--value">
      <div class="Select-value State-SA">
        <span class="Select-value-label" role="option" aria-selected="true" id="react-select-2--value-item">South Australia</span>
      </div>
      <div class="Select-input" style="display: inline-block;">
        <input id="state-select" aria-activedescendant="react-select-2--value" aria-expanded="false" aria-haspopup="false" aria-owns="" role="combobox" value="" style="box-sizing: content-box; width: 5px;">
        <div style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: 14px; font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-weight: 400; font-style: normal; letter-spacing: normal; text-transform: none;">
        </div>
      </div>
    </span>
    <span aria-label="Clear value" class="Select-clear-zone" title="Clear value">
    <span class="Select-clear">×</span></span><span class="Select-arrow-zone"><span class="Select-arrow"></span></span>
  </div>
  <div class="Select-menu-outer">
    <div class="Select-menu" id="react-select-2--list" role="listbox" tabindex="-1">
      <div class="Select-option State-ACT" role="option" aria-label="Australian Capital Territory"
           id="react-select-2--option-0">Australian Capital Territory
      </div>
      <div class="Select-option State-NSW" role="option" aria-label="New South Wales" id="react-select-2--option-1">
        New South Wales
      </div>
      <div class="Select-option State-Vic" role="option" aria-label="Victoria" id="react-select-2--option-2">
        Victoria
      </div>
      <div class="Select-option State-Qld" role="option" aria-label="Queensland" id="react-select-2--option-3">
        Queensland
      </div>
      <div class="Select-option State-WA" role="option" aria-label="Western Australia"
           id="react-select-2--option-4">Western Australia
      </div>
      <div class="Select-option is-selected State-SA" role="option" aria-label="South Australia"
           id="react-select-2--option-5">South Australia
      </div>
      <div class="Select-option State-Tas" role="option" aria-label="Tasmania" id="react-select-2--option-6">
        Tasmania
      </div>
      <div class="Select-option is-focused State-NT" role="option" aria-label="Northern Territory"
           id="react-select-2--option-7">Northern Territory
      </div>
    </div>
  </div>
</div>
```

**Comment by techiemon** [comment]
I used the valueComponent and menuRenderer options with these methods to get the focused option and selected value to be spoken by VoiceOver. 

`
const ariaStyle: any = {
  position: 'absolute',
  width: '1px',
  height: '1px',
  padding: '0',
  margin: '-1px',
  overflow: 'hidden',
  clip: 'rect(0, 0, 0, 0)',
  border: 0,
};

const menuRenderer = params => {
  return (
    <div className="Select-menu-outer">
      <span role="status"
            aria-live="assertive"
            aria-atomic="true"
            style={ariaStyle}
      >
        {params.focusedOption.label}. {params.options.length} list items available, press enter to choose or use arrows to select another option.
      </span>
      <div className="Select-menu" role="listbox">
        {params.options.map(option => {
          return <div key={option.value}
                    onClick={() => params.selectValue(option)}
                    onMouseOver={() => params.focusOption(option)}
                    className={`Select-option ${params.focusedOption && params.focusedOption.value === option.value ? 'is-focused' : ''}`}
                    role="option">{option.label}</div>;
        })}
      </div>
    </div>
  );
};

const valueComponent = params => {
  return (
    <div className="Select-value" title={params.value.label}>
      <span role="status"
            aria-live="assertive"
            aria-atomic="true"
            style={ariaStyle}
      >Selected value, {params.children}
      </span>
      <span className="Select-value-label">
        {params.children}
      </span>
    </div>
  );
};`

**Comment by EnshaednHiker** [comment]
@techiemon, could you post an entire working example of this this code working with a react component? This could be exactly what my company needs for a project.

**Comment by tarikhamilton** [comment]
Also, if anyone could outline what exactly is needed to make this screen reader accessible, I'd love to make a PR. Is @techiemon 's solution adequate?

**Comment by tomasdev** [comment]
Not really, it's more a workaround than the proper solution.

1. ARIA roles need to be added for option, listbox and combobox depending on the case.
2. ARIA attributes for selected, activedescendant, expanded, label are required as well.
3. Keyboard navigation must be validated against WCAG standards (i.e., screen readers must not collide with keyboard navigation)

But then when using autocomplete and AJAX loading, it becomes a little harder.

**Comment by bladey** [comment]
Hello -

In an effort to sustain the `react-select` project going forward, we're closing old issues.

We understand this might be inconvenient but in the best interest of supporting the broader community we have to direct our efforts towards the current major version.

If you aren't using the latest version of `react-select` please consider upgrading to see if it resolves any issues you're having.

However, if you feel this issue  is still relevant and you'd like us to review it - please leave a comment and we'll do our best to get back to you!
